Test Functionality

brennan-obc opened the pull request:
HTML parser added, matchID extraction testing successful

brennan-obc committed:
-added Cheerio HTML Parser as dependency
-removed "@types/axios" from devDependencies

-getMatchIds()
-added html? as optional parameter, allowing HTML to be injected from within for testing purposes
  -in normal operation, function will pull data from axios call
    * test file must be neutralized--if test file is present it will be used
  -when testing, function will pull data from test-file
-added Cheerio html parser
  -returns 'matchIds' array, containg all match IDs from the given input

brennan-obc commented:
# test functionality added
- added Cheerio HTML Parser as dependency
- removed "@types/axios" from devDependencies

## getMatchIds()
- added html? as optional parameter, allowing HTML to be injected from within for testing purposes
  - function can stay as-is in normal operation as long as test file is neutralized
-  added Cheerio html parser
  -  returns 'matchIds' array, containg all match IDs from the given input

briansimoni commented on src/lib/crawler/crawler.ts:
```diff
@@ -1,19 +1,47 @@
-
+// implementing the Crawler interface in the WarmaneCrawler class
 export class WarmaneCrawler implements Crawler {
+  // defining 'getMatchIds' method with optional paramater for testing (html)
   async getMatchIds(params: {
     character: string;
     realm: string;
+    // added optional paramater (HTML passed as string) for testing
+    html?: string;
```
So generally we want to avoid having optional parameters in function signatures just for testing purposes. Your code does actually work though. There was a defect in my unit test which probably caused a lot of your confusion and changes.

When you use the `html` parameter when calling `getMatchIds()`, it won't use the axios call at all, therefore it is not tested at all. You can see this in code coverage tools.

![image](https://user-images.githubusercontent.com/11894935/228065711-84484285-86fa-4c9d-9198-1e149c89a5c0.png)

It's kind of hard to make sense of that at first glance but the line is yellow is indicating a partial test coverage i.e. one side of the nullish coalessence operator is tested but the other is not.

The red line indicates that there is no test coverage. In my opinion, I think it should've highlighted line 21 through 25 as red because it's all one expression, but JavaScript and it's tooling is weird sometimes.

The defect in my unit test was that I was returning a string in my mock axios implementation where I should've been returning an axios response object with a property data and string value.

It was:
```TypeScript
    mockedAxios.get.mockImplementation(async () => testHtml);
```

when it should've been

```TypeScript
    mockedAxios.get.mockImplementation(async () => ({
      data: testHtml,
    }));
```

I request that we remove the optional `html` parameter and correct the mock implementation.

Finally, we have happened upon some inconvenient timing. Apparently the arena season ended on warmane and the data coming back no longer matches what is in my test file. This was inevitable, but it is kind of inconvenient for short term testing purposes. Where I had 453 games played, it now says I have 1.

briansimoni commented on src/lib/crawler/crawler.ts:
```diff
@@ -1,19 +1,47 @@
+    // getMatchId will use optional parameter if present, will only use URL in its absence
+    const html =
+      params.html ??
+      (await axios
+        .get(
+          `https://armory.warmane.com/character/${params.character}/${params.realm}/match-history`
+        )
+        .then((response) => response.data));
```
So this works, but I personally prefer to not combine `await` `.then` `.catch` or `.finally`, in one expression. 

To be honest, with the advent of `async/await` I actually try to never use `.then` `.catch` or `.finally` in production code.

I recommend:
```TypeScript
  const response = await axios.get<string>(
    `https://armory.warmane.com/character/${params.character}/${params.realm}/match-history`
  );
  const html: string = response.data
```
This is TypeScript thing that might be new to you but the `string` part in  `axios.get<string>()`, is a type parameter. It's basically saying, "The data property on the response object will be of type `string`"

briansimoni commented on src/lib/crawler/crawler.ts:
```diff
@@ -1,19 +1,47 @@
+
+    // iterating through each table row with an ID of 'data-table-history'
+    //  - _index not used in callback, but needed as argument (index of current element, current element)
+    $("table#data-table-history tbody tr").each((_index, element) => {
+      // extracting 1st table cell's text, trimming whitespace, adding to 'matchIds' array
+      const matchId = $(element).find("td:first-child").text().trim();
+      // adding extracted ID to array
+      matchIds.push(matchId);
```
👍

briansimoni reviewed (changes requested):
Congratulations on what I assume is your very first pull request and first code review performed by a professional programmer!

You figured out how to fork the entire repository and PR from your own repo. This works and is common in the open source community, but I think you have direct access to push branches to the repository that I'm using. You just can't update the `v2` branch directly because I want to review your code before merging stuff into that branch. I'm fine with either technique for now. Having a forked repository on your profile might be good for your resume anyways.

Your code does work, but I'm requesting a couple of changes before it's merged. See the comments.

briansimoni commented:
Oh and you need to commit `package.json` and `package.lock.json`.

brennan-obc committed:
crawler.ts
-Removed optional 'html' variable
-Replaced line with nullish coalescing operator to seperate lines that will fetch HTML w/ axios and extract content

crawler.test.ts
-imported axios
-created a mock axios
-updated test to mock axios implementation (returns object w/ data prop containing 'testHtml' string instead of directly returning 'testHtml' string)
-removed 'html' parameter when calling getMatchIds() as no longer necessary

briansimoni reviewed (approved):
Good stuff. Approved!
